hazelcast/src/test/java/com/hazelcast/cp/CPGroupAvailabilityListenerTest.java

```diff
@@ -135,4 +140,75 @@ public void majorityLost(CPGroupAvailabilityEvent event) {
+        private final AtomicInteger majorityLost;
+
+        private final Set<UUID> membersShutdown;
+
+        GracefulShutdownAvailabilityListener() {
+            availabilityDecreased = new AtomicInteger();
+            majorityLost = new AtomicInteger();
+            membersShutdown = Collections.synchronizedSet(new HashSet<>());
+        }
+
+        @Override
+        public void availabilityDecreased(CPGroupAvailabilityEvent event) {
+            handleEvent(event, true);
+        }
+
+        @Override
+        public void majorityLost(CPGroupAvailabilityEvent event) {
+            handleEvent(event, false);
+        }
+
+        private void handleEvent(CPGroupAvailabilityEvent event, boolean isAvailabilityDecrease) {
+            if (event instanceof CPGroupAvailabilityEventGracefulImpl) {
+                membersShutdown.addAll(event.getUnavailableMembers().stream().map(CPMember::getUuid).collect(Collectors.toSet()));
+                if (isAvailabilityDecrease) {
+                    availabilityDecreased.incrementAndGet();
+                } else {
+                    majorityLost.incrementAndGet();
+                }
+            }
+        }
+    }
+
+    @Test
+    public void whenMemberShutdown_thenReceiveGracefulAvailabilityEvents_3_cpNodeCount() {
+        whenMemberShutdown_thenReceiveGracefulAvailabilityEvents(3);
+    }
+
+    @Test
+    public void whenMemberShutdown_thenReceiveGracefulAvailabilityEvents_5_cpNodeCount() {
+        whenMemberShutdown_thenReceiveGracefulAvailabilityEvents(5);
+    }
+
+    public void whenMemberShutdown_thenReceiveGracefulAvailabilityEvents(int cpNodeCount) {
+        GracefulShutdownAvailabilityListener listener = new GracefulShutdownAvailabilityListener();
+
+        HazelcastInstance[] instances = newInstances(cpNodeCount);
+        // instances[0] is the only instance we won't shut down; it will observe all the applicable availability events
+        instances[0].getCPSubsystem().addGroupAvailabilityListener(listener);
+
+        HashSet<UUID> expectedShutdownInstanceUuids = new HashSet<>();
+        for (int i = 1; i < instances.length; i++) {
+            expectedShutdownInstanceUuids.add(instances[i].getLocalEndpoint().getUuid());
+        }
+
+        // shutdown the other members until we have only two instances left:
+        //   - instances[0]                   : the one with our event handler attached
+        //   - instances[instances.length - 1]: we'll shut down his one separately as it will result in loss of majority
+        for (int i = 1; i < instances.length - 1; i++) {
+            instances[i].getLifecycleService().shutdown();
+            assertEqualsEventually(i, listener.availabilityDecreased);
+        }
+
+        instances[instances.length - 1].getLifecycleService().shutdown();
+        assertEqualsEventually(1, listener.majorityLost);
+
+        assertEquals(expectedShutdownInstanceUuids, listener.membersShutdown);
+    }
```

Comment:
It would be nice to add a test for the following scenario, in the case of `getLifecycleService().terminate()`.

Let's consider the 3 CP members cluster with non-graceful shutdown:

- 3rd member has been shutdown -> the `availabilityDecreased` event is published
- after the `missing-cp-member-auto-removal-seconds` the 3rd member is auto-removed, the CP group shrink and the majority becomes equal 2, the `memberRemoved` event is published
- 2nd member shutdown -> the `availabilityDecreased` event SHOULD be published

Reply:
even though it's not an issue with this PR, it seems CP availability listener is lacking in terms of test coverage. e.g. I don't see any test where `clusteMemberCount > cpNodeCount > groupSize` to ensure no availability event is triggered when a non-CP member is terminated/shutdown (or any tests where multiple CPGroups are involved).